src/crypto/openssl.rs

```diff
@@ -0,0 +1,941 @@
+                verifier
+                    .verify_oneshot(signature, signed_data)
+                    .map_err(|_| AlgorithmError::InvalidData)?
+            }
+            PublicKey::EcDsa(digest_algorithm, public_key, _) => {
+                let half_len = signature.len() / 2;
+                let mut hasher = Hasher::new(*digest_algorithm)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                hasher
+                    .update(signed_data)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let hash = hasher
+                    .finish()
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let r = BigNum::from_slice(&signature[0..half_len])
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let s = BigNum::from_slice(&signature[half_len..])
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let ecdsa_sig = EcdsaSig::from_private_components(r, s)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                ecdsa_sig
+                    .verify(hash.as_ref(), public_key)
+                    .map_err(|_| AlgorithmError::InvalidData)?
+            }
+        };
+        if valid {
+            Ok(())
+        } else {
+            Err(AlgorithmError::BadSig)
+        }
+    }
+
+    /// Convert to a [`Dnskey`].
+    pub fn dnskey(&self) -> Dnskey<Vec<u8>> {
+        match self {
+            PublicKey::Default(message_digest, public_key, flags) => {
+                let alg = if *message_digest == MessageDigest::sha256() {
+                    SecAlg::RSASHA256
+                } else {
+                    unreachable!();
+                };
+                let rsa = public_key.rsa().expect("should not fail");
+                let e = rsa.e().to_vec();
+                let n = rsa.n().to_vec();
+
+                let mut key = Vec::new();
+
+                // Encode the exponent length.
+                if let Ok(exp_len) = u8::try_from(e.len()) {
+                    key.reserve_exact(1 + e.len() + n.len());
+                    key.push(exp_len);
+                } else if let Ok(exp_len) = u16::try_from(e.len()) {
+                    key.reserve_exact(3 + e.len() + n.len());
+                    key.push(0u8);
+                    key.extend(&exp_len.to_be_bytes());
+                } else {
+                    unreachable!(
+                        "RSA exponents are (much) shorter than 64KiB"
+                    )
+                }
+
+                key.extend(&*e);
+                key.extend(&*n);
+
+                Dnskey::new(*flags, 3, alg, key).expect("should not fail")
+            }
+            PublicKey::NoDigest(public_key, flags) => {
+                let alg = match public_key.id() {
+                    Id::ED25519 => SecAlg::ED25519,
+                    Id::ED448 => SecAlg::ED448,
+                    _ => todo!(),
```

Comment:
As no variants other than ED25519 and ED448 are supported elsewhere in this code, perhaps this should be `unreachable!()` instead of `todo!()` ?